### PR TITLE
[YUNIKORN-1886] Optimize scheduling cycle if there's no placeholder allocation or reservation

### PR DESCRIPTION
**Pull request by pbacsko**
### What is this PR for?
Count reservations and active placeholder allocations so that we can avoid calling allocation code paths unnecessarily.

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-1886

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

File: pkg/scheduler/partition_test.go [hunk 1]
```diff
@@ -2513,6 +2513,7 @@ func TestUpdateNode(t *testing.T) {
 }
 
 func TestAddTGApplication(t *testing.T) {
+
```
Comment (wilfred-s) [inline pkg/scheduler/partition_test.go:2516]: unexpected newline

File: pkg/scheduler/partition.go [hunk 2]
```diff
@@ -567,6 +569,9 @@ func (pc *PartitionContext) AddNode(node *objects.Node, existingAllocations []*o
 				metrics.GetSchedulerMetrics().IncFailedNodes()
 				return err
 			}
+			if alloc.IsPlaceholder() {
+				pc.incPhAllocationCount()
+			}
```
Comment (wilfred-s) [inline pkg/scheduler/partition.go:574]: This one needs to be reversed also if the node is removed.
The node removal path is not currently covered: see line 698 (after change) or 693 (before change) need a decrement placeholder count in that check

Comment (pbacsko) [reply pkg/scheduler/partition.go]: Now we add and remove in two methods:
* `removeNodeAllocations()` / `removeAllocation()` (one/ all when uuid = "")
* `allocate()` (normal scheduling) / `addAllocation()` (startup)

**Review by manirajv06 — APPROVED** [review]
looks good

**Review by wilfred-s — APPROVED** [review]
LGTM

**Pull request closed without merging**